lib/IRGen/GenValueWitness.cpp

```diff
@@ -987,10 +987,19 @@ getAddrOfKnownValueWitnessTable(IRGenModule &IGM, CanType type) {
     // the GOT.
     if (nom->isGenericContext())
       return nullptr;
-    // TODO: Enums need additional value witnesses for their tag manipulation.
-    if (isa<EnumDecl>(nom))
-      return nullptr;
+    // TODO: Non-C enums need additional value witnesses for their tag
+    // manipulation.
+    if (auto enumDecl = dyn_cast<EnumDecl>(nom))
+      if (!enumDecl->isObjC() && !type->isStructurallyUninhabited())
+        return nullptr;
   }
+ 
+  auto &C = IGM.Context;
+
+  // It doesn't really matter what value witnesses uninhabited types use since
+  // they can never be instantiatied. Use an empty type.
```

Comment:
`isStructurallyInhabited()` seems to include tuples that contain structurally-inhabited types, and those tuples won't have size 0 unless we add that as a special-case rule (which is probably not worth doing).

Maybe this could just be a general rule about types with size 0?

Reply:
A tuple with structurally-uninhabited members cannot be laid out in memory, so it may as well have size 0.

Reply:
It can't be *occupied* in memory.  It can still be a part of a layout, e.g. in an impossible enum case, and unless we have a layout rule that actually changes its size to 0 we need to maintain consistency.